File: DKImagePickerControllerDemo_OC/DKImagePickerControllerDemo_OC/Source/DKImagePickerController/View/DKAssetGroupListVC.m
//
//  DKAssetGroupListVC.m
//  DKImagePickerControllerDemo_OC
//
//  Created by 赵铭 on 2017/6/29.
//  Copyright © 2017年 zm. All rights reserved.
//

#import "DKAssetGroupListVC.h"
#import "DKImageResource.h"
#import "DKImageManager.h"
#import "DKGroupDataManager.h"
#import "DKAssetGroup.h"
#import "DKPopoverViewController.h"
static NSString * DKImageGroupCellIdentifier = @"DKImageGroupCellIdentifier";

@interface DKAssetGroupSeparator : UIView


@end

@implementation DKAssetGroupSeparator
- (UIColor*)backgroundColor{
    return [super backgroundColor];
}
- (void)setBackgroundColor:(UIColor *)backgroundColor{
    if (backgroundColor != [UIColor clearColor]) {
        super.backgroundColor = backgroundColor;
    }
}
@end

@interface DKAssetGroupCell : UITableViewCell
@property (nonatomic, strong) UIImageView * thumbnailImageView;
@property (nonatomic, strong) UILabel * groupNameLabel;
@property (nonatomic, strong) UILabel * totalCountLabel;
@property (nonatomic, strong) UIView * customSelectedBackgroundView;
@property (nonatomic, strong) DKAssetGroupSeparator * customSeparator;
@end

@implementation DKAssetGroupCell

- (UIImageView *)thumbnailImageView{
    if (!_thumbnailImageView) {
        _thumbnailImageView = [UIImageView new];
        _thumbnailImageView.contentMode = UIViewContentModeScaleAspectFill;
        _thumbnailImageView.clipsToBounds = YES;
    }
    return _thumbnailImageView;
}

- (UILabel *)groupNameLabel{
    if (!_groupNameLabel) {
        _groupNameLabel = [UILabel new];
        _groupNameLabel.font = [UIFont boldSystemFontOfSize:13];
    }
    return _groupNameLabel;
}

- (UILabel *)totalCountLabel{
    if (!_totalCountLabel) {
        _totalCountLabel = [UILabel new];
        _totalCountLabel.font = [UIFont systemFontOfSize:10];
        _totalCountLabel.textColor = [UIColor grayColor];
    }
    return _totalCountLabel;
}

- (UIView *)customSelectedBackgroundView{
    if (!_customSelectedBackgroundView) {
        _customSelectedBackgroundView = [UIView new];
        UIImageView * selectedFlag = [[UIImageView alloc] initWithImage:[DKImageResource blueTickImage]];
        selectedFlag.frame = CGRectMake(_customSelectedBackgroundView.bounds.size.width - selectedFlag.bounds.size.width - 20,(_customSelectedBackgroundView.bounds.size.width - selectedFlag.bounds.size.width) / 2, selectedFlag.bounds.size.width, selectedFlag.bounds.size.height);
        selectedFlag.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin;
        [_customSelectedBackgroundView addSubview:selectedFlag];
    }
    
    return _customSelectedBackgroundView;
}

- (DKAssetGroupSeparator *)customSeparator{
    if (!_customSeparator ) {
        _customSeparator = [[DKAssetGroupSeparator alloc] initWithFrame:CGRectMake(10, self.bounds.size.height - 1, self.bounds.size.width, 0.5)];
        _customSeparator.backgroundColor = [UIColor lightGrayColor];
        _customSeparator.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleTopMargin;
    }
    return _customSeparator;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        self.selectedBackgroundView = self.customSelectedBackgroundView;
        [self.contentView addSubview:self.thumbnailImageView];
        [self.contentView addSubview:self.groupNameLabel];
        [self.contentView addSubview:self.totalCountLabel];
        [self addSubview:self.customSeparator];
    }
    
    return self;
}

- (void)layoutSubviews{
    [super layoutSubviews];
    CGFloat imageViewY = 10.0;
    CGFloat imageViewHeight = self.contentView.bounds.size.height - 2*imageViewY;
    self.thumbnailImageView.frame = CGRectMake(imageViewY, imageViewY, imageViewHeight, imageViewHeight);
    
    self.groupNameLabel.frame = CGRectMake(CGRectGetMaxX(self.thumbnailImageView.frame) + 10, CGRectGetMinY(self.thumbnailImageView.frame) + 5 , 200, 20);
    
    self.totalCountLabel.frame = CGRectMake(CGRectGetMinX(self.groupNameLabel.frame), CGRectGetMaxY(self.thumbnailImageView.frame) - 20, 200, 20);
    
}

@end





@interface DKAssetGroupListVC ()
@property (nonatomic, copy) NSArray <NSString *>*groups;
@property (nonatomic, copy) NSString * selectedGroup;

@property (nonatomic, strong)PHImageRequestOptions *groupThumbnailRequestOptions;
@end

@implementation DKAssetGroupListVC
- (instancetype)initWithSelectedGroupDidChangeBlock:(void(^)(NSString * groupId))selectedGroupDidChangeBlock
                                  defaultAssetGroup:(PHAssetCollectionSubtype)defaultAssetGroup{
    if (self = [super initWithStyle:UITableViewStylePlain]) {
        self.defaultAssetGroup = defaultAssetGroup;
        self.selectedGroupDidChangeBlock = selectedGroupDidChangeBlock;
    }
    return self;
}

- (PHImageRequestOptions *)groupThumbnailRequestOptions{
    if (!_groupThumbnailRequestOptions) {
        _groupThumbnailRequestOptions = [PHImageRequestOptions new];
        _groupThumbnailRequestOptions.deliveryMode = PHImageRequestOptionsDeliveryModeOpportunistic;
        _groupThumbnailRequestOptions.resizeMode = PHImageRequestOptionsResizeModeExact;
    }
    return _groupThumbnailRequestOptions;
}
- (CGSize)preferredContentSize{
    if (self.groups.count > 0) {
        return CGSizeMake(UIViewNoIntrinsicMetric, _groups.count * self.tableView.rowHeight);
    }else{
        return [super preferredContentSize];
    }
}
- (void)setPreferredContentSize:(CGSize)preferredContentSize{
    super.preferredContentSize = preferredContentSize;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self .tableView registerClass:[DKAssetGroupCell class] forCellReuseIdentifier:DKImageGroupCellIdentifier];
    self.tableView.rowHeight = 70;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}
- (void)loadGroups{
    
    __weak typeof(self) weakSelf = self;
    [[[DKImageManager shareInstance] groupDataManager] fetchGroupsWithCompleteBlock:^(NSArray<NSString *> *groupIds, NSError *error) {
        __strong typeof(self) strongSelf = weakSelf;
        strongSelf.groups = groupIds;
        strongSelf.selectedGroup = [strongSelf defaultAssetGroupOfAppropriate];
        if (strongSelf.selectedGroup) {
            NSInteger row = [strongSelf.groups indexOfObject:strongSelf.selectedGroup];
            
            [strongSelf.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:row inSection:0] animated:YES scrollPosition:UITableViewScrollPositionNone];
            
        }
        if (strongSelf.selectedGroupDidChangeBlock) {
            strongSelf.selectedGroupDidChangeBlock(strongSelf.selectedGroup);
        }
    }];
}

- (NSString *)defaultAssetGroupOfAppropriate{
    if (!self.groups) {
        return nil;
    }
    for (NSString * groupId in self.groups) {
       DKAssetGroup * group = [[[DKImageManager shareInstance] groupDataManager] fetchGroupWithGroupId:groupId
         ];
        if (self.defaultAssetGroup == group.originalCollection.assetCollectionSubtype) {
            return groupId;
        }
    
    }
    return self.groups.firstObject;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.groups.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    DKAssetGroupCell *cell = [tableView dequeueReusableCellWithIdentifier:DKImageGroupCellIdentifier forIndexPath:indexPath];
    
    DKAssetGroup * group = [[[DKImageManager shareInstance] groupDataManager] fetchGroupWithGroupId:self.groups[indexPath.row]];
    cell.groupNameLabel.text = group.groupName;
    NSInteger tag = indexPath.row + 1;
    cell.tag = tag;
    
    if (group.totalCount == 0) {
        cell.thumbnailImageView.image= [DKImageResource emptyAlbumIcon];
    }else{
        [[[DKImageManager shareInstance] groupDataManager] fetchGroupThumbnailForGroup:group.groupId size:CGSizeMake(tableView.rowHeight, tableView.rowHeight) options:self.groupThumbnailRequestOptions completeBlock:^(UIImage *image, NSDictionary *info) {
            if (cell.tag == tag) {
                cell.thumbnailImageView.image = image;
            }
        }];
    }
    
    cell.totalCountLabel.text = [NSString stringWithFormat:@"%ld", (long)group.totalCount];
    return cell;
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [DKPopoverViewController dismissPopoverViewController];
    self.selectedGroup = _groups[indexPath.row];
    self.selectedGroupDidChangeBlock(self.selectedGroup);
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: DKImagePickerController/View/DKAssetGroupListVC.swift
//
//  DKAssetGroupListVC.swift
//  DKImagePickerController
//
//  Created by ZhangAo on 15/8/8.
//  Copyright (c) 2015年 ZhangAo. All rights reserved.
//

import Photos

let DKImageGroupCellIdentifier = "DKImageGroupCellIdentifier"

class DKAssetGroupCell: UITableViewCell {

    class DKAssetGroupSeparator: UIView {

        override var backgroundColor: UIColor? {
            get {
                return super.backgroundColor
            }
            set {
                if newValue != UIColor.clear {
                    super.backgroundColor = newValue
                }
            }
        }
    }

    fileprivate let thumbnailImageView: UIImageView = {
        let thumbnailImageView = UIImageView()
        thumbnailImageView.contentMode = .scaleAspectFill
        thumbnailImageView.clipsToBounds = true

        return thumbnailImageView
    }()

    var groupNameLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont.boldSystemFont(ofSize: 13)
        return label
    }()

    var totalCountLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 10)
        label.textColor = UIColor.gray
        return label
    }()

    var customSelectedBackgroundView: UIView = {
        let selectedBackgroundView = UIView()

        let selectedFlag = UIImageView(image: DKImageResource.blueTickImage())
        selectedFlag.frame = CGRect(x: selectedBackgroundView.bounds.width - selectedFlag.bounds.width - 20,
                                    y: (selectedBackgroundView.bounds.width - selectedFlag.bounds.width) / 2,
                                    width: selectedFlag.bounds.width, height: selectedFlag.bounds.height)
        selectedFlag.autoresizingMask = [.flexibleLeftMargin, .flexibleTopMargin, .flexibleBottomMargin]
        selectedBackgroundView.addSubview(selectedFlag)

        return selectedBackgroundView
    }()

    lazy var customSeparator: DKAssetGroupSeparator = {
        let separator = DKAssetGroupSeparator(frame: CGRect(x: 10, y: self.bounds.height - 1, width: self.bounds.width, height: 0.5))

        separator.backgroundColor = UIColor.lightGray
        separator.autoresizingMask = [.flexibleWidth, .flexibleTopMargin]
        return separator
    }()

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        self.selectedBackgroundView = self.customSelectedBackgroundView

        self.contentView.addSubview(thumbnailImageView)
        self.contentView.addSubview(groupNameLabel)
        self.contentView.addSubview(totalCountLabel)

        self.addSubview(self.customSeparator)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        let imageViewY = CGFloat(10)
        let imageViewHeight = self.contentView.bounds.height - 2 * imageViewY
        self.thumbnailImageView.frame = CGRect(x: imageViewY,
                                               y: imageViewY,
                                               width: imageViewHeight,
                                               height: imageViewHeight)

        self.groupNameLabel.frame = CGRect(
            x: self.thumbnailImageView.frame.maxX + 10,
            y: self.thumbnailImageView.frame.minY + 5,
            width: 200,
            height: 20)

        self.totalCountLabel.frame = CGRect(
            x: self.groupNameLabel.frame.minX,
            y: self.thumbnailImageView.frame.maxY - 20,
            width: 200,
            height: 20)
    }
}

class DKAssetGroupListVC: UITableViewController, DKGroupDataManagerObserver {

    convenience init(selectedGroupDidChangeBlock: @escaping (_ groupId: String?) -> (), defaultAssetGroup: PHAssetCollectionSubtype?) {
        self.init(style: .plain)

        self.defaultAssetGroup = defaultAssetGroup
        self.selectedGroupDidChangeBlock = selectedGroupDidChangeBlock
    }

    fileprivate var groups: [String]?

    fileprivate var selectedGroup: String?

    fileprivate var defaultAssetGroup: PHAssetCollectionSubtype?

    fileprivate var selectedGroupDidChangeBlock:((_ group: String?)->())?

    fileprivate lazy var groupThumbnailRequestOptions: PHImageRequestOptions = {
        let options = PHImageRequestOptions()
        options.deliveryMode = .opportunistic
        options.resizeMode = .exact

        return options
    }()

    override var preferredContentSize: CGSize {
        get {
            if let groups = self.groups {
                return CGSize(width: UIViewNoIntrinsicMetric,
                              height: CGFloat(groups.count) * self.tableView.rowHeight)
            } else {
                return super.preferredContentSize
            }
        }
        set {
            super.preferredContentSize = newValue
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.tableView.register(DKAssetGroupCell.self, forCellReuseIdentifier: DKImageGroupCellIdentifier)
        self.tableView.rowHeight = 70
        self.tableView.separatorStyle = .none

        self.clearsSelectionOnViewWillAppear = false

        getImageManager().groupDataManager.addObserver(self)
    }

    internal func loadGroups() {
        getImageManager().groupDataManager.fetchGroups { [weak self] groups, error in
            guard let groups = groups, let strongSelf = self, error == nil else { return }

            strongSelf.groups = groups
            strongSelf.selectedGroup = strongSelf.defaultAssetGroupOfAppropriate()
            if let selectedGroup = strongSelf.selectedGroup,
                let row =  groups.index(of: selectedGroup)
            {
                strongSelf.tableView.selectRow(
                    at: IndexPath(row: row, section: 0),
                    animated: false,
                    scrollPosition: .none)
            }

            strongSelf.selectedGroupDidChangeBlock?(strongSelf.selectedGroup)
        }
    }

    fileprivate func defaultAssetGroupOfAppropriate() -> String? {
        guard let groups = self.groups else { return nil }
        if let defaultAssetGroup = self.defaultAssetGroup {
            for groupId in groups {
                let group = getImageManager().groupDataManager.fetchGroupWithGroupId(groupId)
                if defaultAssetGroup == group.originalCollection.assetCollectionSubtype {
                    return groupId
                }
            }
        }
        return groups.first
    }

    // MARK: - UITableViewDelegate, UITableViewDataSource methods

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return groups?.count ?? 0
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let groups = groups,
            let cell = tableView.dequeueReusableCell(
                withIdentifier: DKImageGroupCellIdentifier,
                for: indexPath) as? DKAssetGroupCell else
        {
            assertionFailure("Expect groups and cell")
            return UITableViewCell()
        }

        let assetGroup = getImageManager().groupDataManager.fetchGroupWithGroupId(groups[indexPath.row])
        cell.groupNameLabel.text = assetGroup.groupName

        let tag = indexPath.row + 1
        cell.tag = tag

        if assetGroup.totalCount == 0 {
            cell.thumbnailImageView.image = DKImageResource.emptyAlbumIcon()
        } else {
            getImageManager().groupDataManager.fetchGroupThumbnailForGroup(
                assetGroup.groupId,
                size: CGSize(width: tableView.rowHeight, height: tableView.rowHeight).toPixel(),
                options: self.groupThumbnailRequestOptions) { image, info in
                    if cell.tag == tag {
                        cell.thumbnailImageView.image = image
                    }
            }
        }
        cell.totalCountLabel.text = String(assetGroup.totalCount)

        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        DKPopoverViewController.dismissPopoverViewController()
        
        guard let groups = self.groups, groups.count > indexPath.row else {
            assertionFailure("Expect groups with count > \(indexPath.row)")
            return
        }
        
        self.selectedGroup = groups[indexPath.row]
        selectedGroupDidChangeBlock?(self.selectedGroup)
    }
    
    // MARK: - DKGroupDataManagerObserver methods
    
    func groupDidUpdate(_ groupId: String) {
        self.tableView.reloadData()
    }
    
    func groupsDidInsert(_ groupIds: [String]) {
        self.groups! += groupIds
        
        self.willChangeValue(forKey: "preferredContentSize")
        
        let indexPathForSelectedRow = self.tableView.indexPathForSelectedRow
        self.tableView.reloadData()
        self.tableView.selectRow(at: indexPathForSelectedRow, animated: false, scrollPosition: .none)
        
        self.didChangeValue(forKey: "preferredContentSize")
    }
    
    func groupDidRemove(_ groupId: String) {
        guard let row = self.groups?.index(of: groupId) else { return }
        
        self.willChangeValue(forKey: "preferredContentSize")
        
        self.groups?.remove(at: row)
        
        self.tableView.reloadData()
        if self.selectedGroup == groupId {
            self.selectedGroup = self.groups?.first
            selectedGroupDidChangeBlock?(self.selectedGroup)
            self.tableView.selectRow(at: IndexPath(row: 0, section: 0),
                                     animated: false,
                                     scrollPosition: .none)
        }
        
        self.didChangeValue(forKey: "preferredContentSize")
    }
}
